cli/command: ps: only check format-template once

thaJeztah opened the pull request:
- do an early check if a custom format is specified either through the
  command-line, or through the cli's configuration, before adjusting
  the options (to add "size" if needed).
- if no format is set after this, set the default (formatter.TableFormatKey)
- also removes a redundant `options.Size = opts.size` line, as this value is
  already copied at the start of buildContainerListOptions()

thaJeztah committed:
cli/command: ps: only check format-template once

- do an early check if a custom format is specified either through the
  command-line, or through the cli's configuration, before adjusting
  the options (to add "size" if needed).
- if no format is set after this, set the default (formatter.TableFormatKey)
- also removes a redundant `options.Size = opts.size` line, as this value is
  already copied at the start of buildContainerListOptions()
- Update NewContainerFormat to use "table" format as a default if no format
  was given.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

photra commented on cli/command/container/list.go:
```diff
@@ -109,33 +108,17 @@ func buildContainerListOptions(opts *psOptions) (*types.ContainerListOptions, er
 func runPs(dockerCli command.Cli, options *psOptions) error {
 	ctx := context.Background()
 
+	if len(options.format) == 0 && !options.quiet {
+		options.format = dockerCli.ConfigFile().PsFormat
+	}
+
```
```suggestion
	if len(options.format) == 0 && !options.quiet {
		if len(dockerCli.ConfigFile().PsFormat) > 0 {
			options.format = dockerCli.ConfigFile().PsFormat
		} else {
			options.format = formatter.TableFormatKey
		}
	}
```

I think we can merge our two if statements into one code block. We initially check whether a format had been specified or if the quiet option had been selected, then proceed to check the length of `PsFormat` within the config file to correctly assign a format.

What do you think about this?

thaJeztah replied:
🤔 let me think (I think I had a reason)

Ah, so we _could_ probably move them into `buildContainerListOptions`. Looking at it, I guess 


The reason I'm not setting the format to `formatter.TableFormatKey` here (but doing that at the end) was for two reasons;

- I wanted the format to stay empty (if no format was given, and no `psFormat` is set) otherwise the template parsing code (for the size check) would be called, but not needed (overall, the code is a bit of a "hack")
- However, if no format is set at the end (even if `quiet` is set), we must set it to the default "table" value, to make this logic work; https://github.com/docker/cli/blob/d0df532a25ad18e8b4325edcf46f3d0f05b47708/cli/command/formatter/container.go#L30-L33


FWIW, I think we need to look at the logic in that code above as well (as a follow-up), to have that code pick the default (if format is empty), but also consider to make  `quiet`  work even if we would pass it a custom format.

thaJeztah replied:
Ah! Right; and the reason I did the `options.format = dockerCli.ConfigFile().PsFormat` here, is that `buildContainerListOptions() didn't have access to `dockerCli` 😅

thaJeztah replied:
I can add a comment to that line though to explain why the logic is the way it is; one moment.

photra replied:
Ah that makes a lot of sense!

photra commented on cli/command/container/list.go:
```diff
@@ -109,33 +108,17 @@ func buildContainerListOptions(opts *psOptions) (*types.ContainerListOptions, er
-		}
+	if len(options.format) == 0 {
+		options.format = formatter.TableFormatKey
 	}
```
```suggestion
```
ditto

thaJeztah commented:
@photra ptal; if you agree with these changes, and merge this PR, it will be included in your PR ☺️👍

photra commented:
Hi @thaJeztah, I just made some suggestions to above for conciseness. Please let me know what you think about it! 😁

thaJeztah commented:
@photra I pushed a change; added one extra commit in which I fixed the logic in `NewContainerFormat` (keeping it as a separate commit for now to make it easier to find back)

thaJeztah commented on cli/command/container/list.go:
```diff
@@ -77,8 +77,7 @@ func buildContainerListOptions(opts *psOptions) (*types.ContainerListOptions, er
 
-	options.Size = opts.size
-	if !options.Size && len(opts.format) > 0 {
+	if !opts.quiet && !options.Size && len(opts.format) > 0 {
```
Moved the `quiet` check here, so that this parsing is never called if `--quiet` is set (in which case it's not needed)

thaJeztah commented:
Actually, perhaps we should do the last commit as a follow-up for visibility; let me remove it for now (and I can push it as a separate PR); possibly it needs discussion as there will be a slight change in behavior, which may need discussion.

I removed the commit, and included some of the change in the existing one 👍

photra commented:
Gotcha, merging this PR now!
